core/network/helpers/message_read_writer.hpp

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_MESSAGE_READ_WRITER_HPP
+#define KAGOME_MESSAGE_READ_WRITER_HPP
+
+#include <functional>
+#include <memory>
+#include <vector>
+#include <gsl/span>
+#include <boost/assert.hpp>
+#include <boost/system/error_code.hpp>
+
+#include "outcome/outcome.hpp"
+
+namespace kagome::network {
+  struct NoSink {};
+
+  /**
+   * Chain specific messages read-writer
+   */
+  template<typename Adapter, typename Ancestor> struct MessageReadWriter final {
+    using AncestorType = Ancestor;
+    using AdapterType = Adapter;
+    using BufferContainer = std::vector<uint8_t>;
+    using SelfType = MessageReadWriter<AdapterType, AncestorType>;
+
+   public:
+    MessageReadWriter() = default;
+    ~MessageReadWriter() = default;
+
+    MessageReadWriter(MessageReadWriter &&) noexcept = default;
+    MessageReadWriter& operator=(MessageReadWriter &&) noexcept = default;
+
+    MessageReadWriter(const MessageReadWriter&) = delete;
+    MessageReadWriter& operator=(const MessageReadWriter&) = delete;
+
+   public:
+    template<typename T>
+    static size_t need_to_reserve(const T &t) {
+      return AdapterType::size(t) + AncestorType::need_to_reserve(t);
+    }
+
+    template<typename T>
+    static BufferContainer::iterator write(const T &t, BufferContainer &out, size_t reserved = 0ull) {
+      const auto need_to_reserve = SelfType::need_to_reserve(t);
+      if (need_to_reserve > out.size())
+        out.resize(need_to_reserve);
+
+      const size_t r = AdapterType::size(t) + reserved;
+      auto loaded = AncestorType::write(t, out, r);
+
+      BOOST_ASSERT(static_cast<size_t>(std::distance(out.begin(), loaded)) >= r);
```

Comment:
Not so sure about the safety of static_cast here. distance may be negative, if something goes wrong in AncestorType::write. I believe it's better to check.

Reply:
It can't. If so it will be a logic error of the code. So I will add an extra ```assertion``` for this.